diff -rupN .rvm/gems/ruby-2.1.2/gems/bud-0.9.7/lib/bud/collections.rb .gem/ruby/2.1.0/gems/bud-0.9.7/lib/bud/collections.rb
--- .rvm/gems/ruby-2.1.2/gems/bud-0.9.7/lib/bud/collections.rb	2014-09-16 14:13:34.000000000 -0400
+++ .gem/ruby/2.1.0/gems/bud-0.9.7/lib/bud/collections.rb	2015-01-21 15:18:16.000000000 -0500
@@ -520,6 +520,21 @@ module Bud
       end
     end
 
+    def send_tcp_data(the_locspec)
+      establish_connection(the_locspec) if toplevel.connections[the_locspec].nil?
+      # if the connection failed, we silently ignore and let the tuples be cleared.
+      # if we didn't clear them here, we'd be clearing them at end-of-tick anyhow
+      unless toplevel.connections[the_locspec].nil?
+          toplevel.connections[the_locspec].send_data @wire_buf.string
+      end
+    end
+
+    def establish_connection(l)
+      toplevel = @bud_instance.toplevel
+      toplevel.connections[l] = EventMachine::connect l[0], l[1], BudServer, @bud_instance
+      toplevel.connections.delete(l) if toplevel.connections[l].error?
+    end
+
     # Assign self a schema, by hook or by crook.  If +o+ is schemaless *and*
     # empty, will leave @cols as is.
     private
@@ -1010,9 +1025,12 @@ module Bud
         end
         @packer.write(marshall_indexes)
         @packer.flush
-        toplevel.dsock.send_datagram(@wire_buf.string,
-                                     the_locspec[0], the_locspec[1])
 
+        if toplevel.options[:tcp] 
+          send_tcp_data(the_locspec)
+        else 
+          toplevel.dsock.send_datagram(@wire_buf.string, the_locspec[0], the_locspec[1])
+        end
         # Reset output buffer
         @wire_buf.rewind
         @wire_buf.truncate(0)
diff -rupN .rvm/gems/ruby-2.1.2/gems/bud-0.9.7/lib/bud/server.rb .gem/ruby/2.1.0/gems/bud-0.9.7/lib/bud/server.rb
--- .rvm/gems/ruby-2.1.2/gems/bud-0.9.7/lib/bud/server.rb	2014-09-16 14:13:34.000000000 -0400
+++ .gem/ruby/2.1.0/gems/bud-0.9.7/lib/bud/server.rb	2014-09-16 23:14:53.000000000 -0400
@@ -1,7 +1,7 @@
 require 'socket'
 
 class Bud::BudServer < EM::Connection #:nodoc: all
-  def initialize(bud, channel_filter)
+  def initialize(bud, channel_filter=nil)
     @bud = bud
     @channel_filter = channel_filter
     @filter_buf = {}
@@ -39,7 +39,7 @@ class Bud::BudServer < EM::Connection #:
       # If we raise an exception here, EM dies, which causes problems (e.g.,
       # other Bud instances in the same process will crash). Ignoring the
       # error isn't best though -- we should do better (#74).
-      puts "Exception handling network messages: #{e}"
+      puts "Exception handling network messages: #{e.class}:#{e}"
       puts e.backtrace
       puts "Inbound messages:"
       @bud.inbound.each do |chn_name, t|
diff -rupN .rvm/gems/ruby-2.1.2/gems/bud-0.9.7/lib/bud.rb .gem/ruby/2.1.0/gems/bud-0.9.7/lib/bud.rb
--- .rvm/gems/ruby-2.1.2/gems/bud-0.9.7/lib/bud.rb	2014-09-16 14:13:34.000000000 -0400
+++ .gem/ruby/2.1.0/gems/bud-0.9.7/lib/bud.rb	2015-01-29 14:33:49.000000000 -0500
@@ -75,6 +75,7 @@ module Bud
   attr_accessor :stratified_rules
   attr_accessor :metrics, :periodics
   attr_accessor :this_rule_context, :qualified_name
+  attr_accessor :connections
   attr_reader :running_async
 
   # options to the Bud runtime are passed in a hash, with the following keys
@@ -143,6 +144,7 @@ module Bud
     @push_sorted_elems = nil
     @running_async = false
     @bud_started = false
+    @connections = {}
 
     # Setup options (named arguments), along with default values
     @options = options.clone
@@ -150,6 +152,7 @@ module Bud
     @ip = @options[:ip]
     @options[:port] ||= 0
     @options[:port] = @options[:port].to_i
+
     # NB: If using an ephemeral port (specified by port = 0), the actual port
     # number won't be known until we start EM
 
@@ -523,7 +526,8 @@ module Bud
       next if dflt_rescan.member? scanner
       next if scanner.class <= LatticeScanner
 
-      rescan = dflt_rescan.clone
+      #VZM rescan = dflt_rescan.clone
+      rescan = dflt_rescan + [scanner]
       invalidate = dflt_invalidate + [scanner.collection]
       rescan_invalidate_tc(stratum, rescan, invalidate)
       prune_rescan_invalidate(rescan, invalidate)
@@ -989,7 +993,11 @@ module Bud
     @timers.each {|t| t.cancel}
     @tables.each_value {|t| t.close}
     if EventMachine::reactor_running? and @bud_started
-      @dsock.close_connection
+      if @options[:tcp] 
+        EventMachine::stop_server @server
+      else
+        @dsock.close_connection
+      end
     end
     @bud_started = false
     @running_async = false
@@ -999,10 +1007,33 @@ module Bud
   end
 
   def do_start_server
-    @dsock = EventMachine::open_datagram_socket(@ip, @options[:port],
-                                                BudServer, self,
-                                                @options[:channel_filter])
-    @port = Socket.unpack_sockaddr_in(@dsock.get_sockname)[0]
+
+    if @options[:tcp] 
+      if @options[:port] == 0
+        success = false
+        15.times do
+          @port = 5000 + rand(20000)
+          begin
+            @server = EventMachine::start_server(@ip, @port, BudServer, self, @options[:channel_filter])
+            success = true
+            @connections = {}
+            break
+          rescue Exception
+            next
+          end
+        end
+        raise "Failed to bind to local TCP port #{@port}" unless success
+      else
+        @port = @options[:port]
+        @server = EventMachine::start_server(@ip, @port, BudServer, self)
+        @connections = {}
+      end
+    else
+      @dsock = EventMachine::open_datagram_socket(@ip, @options[:port],
+                                                  BudServer, self,
+                                                  @options[:channel_filter])
+      @port = Socket.unpack_sockaddr_in(@dsock.get_sockname)[0]
+    end
   end
 
   public
